[quant] Support integer implementations for max_pool2d

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104226
* __->__ #104225

Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
tested by opinfo, python test/test_ops.py

Reviewers:

Subscribers:

Tasks:

Tags:

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

jerryzh168 committed:
[quant] Support integer implementations for max_pool2d

Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
TBD, is it tested by opinfo directly?

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/cpu/MaxPoolKernel.cpp:
```diff
@@ -88,6 +88,42 @@ void cpu_max_pool(
+  return vec.isnan();
+}
+
+template <>
+vec::Vectorized<unsigned char> isnan<unsigned char>(vec::Vectorized<unsigned char> vec) {
+  Vectorized<unsigned char> ret(false);
+  return ret;
+}
+
+template <>
+vec::Vectorized<signed char> isnan<signed char>(vec::Vectorized<signed char> vec) {
+  Vectorized<signed char> ret(false);
+  return ret;
+}
+
+template <>
+vec::Vectorized<short> isnan<short>(vec::Vectorized<short> vec) {
+  Vectorized<short> ret(false);
+  return ret;
+}
+
+template <>
+vec::Vectorized<int> isnan<int>(vec::Vectorized<int> vec) {
+  Vectorized<int> ret(false);
+  return ret;
+}
+
+template <>
+vec::Vectorized<int64_t> isnan<int64_t>(vec::Vectorized<int64_t> vec) {
+  Vectorized<int64_t> ret(false);
+  return ret;
+}
```
Simplify these with SFINAE `std::is_integral` check?

jerryzh168 replied:
good point, thanks

jerryzh168 replied:
looks like it does not work somehow, I'll need to revert this

jgong5 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1810,6 +1810,10 @@ def forward(self, input_tensor, hidden_tensor):
             self.assertEqual(model_fx(*example_inputs), model_graph(*example_inputs))
 
+    def test_maxpool2d_int8(self):
+        x = torch.randint(1, 100, (1, 3, 3, 3), dtype=torch.int8)
```
Is `torch.uint8` a possible input too?

jerryzh168 replied:
yes, this is just a sanity check, I'm not sure where should I add test yet

jerryzh168 committed:
Update on "[quant] Support integer implementations for max_pool2d"


Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
TBD, is it tested by opinfo directly?

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

kimishpatel commented on aten/src/ATen/native/cpu/MaxPoolKernel.cpp:
```diff
@@ -471,7 +481,7 @@ void max_pool2d_kernel_impl(
     }
     case at::MemoryFormat::ChannelsLast: {
-      AT_DISPATCH_FLOATING_TYPES_AND(ScalarType::BFloat16, input.scalar_type(), "max_pool2d_channels_last", [&] {
+      AT_DISPATCH_ALL_TYPES_AND(ScalarType::BFloat16, input.scalar_type(), "max_pool2d_channels_last", [&] {
```
is complex part of ALL_TYPES?

jerryzh168 replied:
no, all_types == floating_point types + integral types: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/Dispatch.h#L378

jerryzh168 committed:
Update on "[quant] Support integer implementations for max_pool2d"


Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
TBD, is it tested by opinfo directly?

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Support integer implementations for max_pool2d"


Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
TBD, is it tested by opinfo directly?

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Support integer implementations for max_pool2d"


Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
TBD, is it tested by opinfo directly?

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Support integer implementations for max_pool2d"


Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
tested by opinfo, python/test/test_ops.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel reviewed (approved):
Looks good. Did you test via python/test/test_ops.py?

jerryzh168 commented:
> Looks good. Did you test via python/test/test_ops.py?

yeah I ran the test locally, there are some test failures unrelated to the PR, maybe they are related to test environment, I think it should be OK to just rely on the OSSCI here

jerryzh168 committed:
Update on "[quant] Support integer implementations for max_pool2d"


Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
tested by opinfo, python test/test_ops.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Support integer implementations for max_pool2d"


Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
tested by opinfo, python test/test_ops.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Support integer implementations for max_pool2d"


Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
tested by opinfo, python test/test_ops.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -447,6 +447,11 @@
         "nn.functional.embedding",
         reason=onnx_test_common.reason_onnx_script_does_not_support("aten.embedding_renorm.default"),
     ),
+    xfail(
```
cc @BowenBao @abock @thiagocrepaldi FYI

jerryzh168 committed:
Update on "[quant] Support integer implementations for max_pool2d"


Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
tested by opinfo, python test/test_ops.py

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Support integer implementations for max_pool2d"


Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
tested by opinfo, python test/test_ops.py

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Support integer implementations for max_pool2d"


Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
tested by opinfo, python test/test_ops.py

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Support integer implementations for max_pool2d"


Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
tested by opinfo, python test/test_ops.py

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Support integer implementations for max_pool2d"


Summary:
This is needed for representing quantized model in pt2 export quantization flow

Test Plan:
tested by opinfo, python test/test_ops.py

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
